keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2862 @@
+admissionregistration.k8s.io group. (suggestions welcome on exact name to use
+for kind)
+
+At a high level, the API will support:
+
+- Request matching (similar to the match rules of admission webhooks, RBAC, priority & fairness and Audit)
+- CEL rule evaluation (similar to both [CRD Validation
+  Rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language)
+  but with access to the data in
+  [AdmissionRequest](https://github.com/kubernetes/kubernetes/blob/2ac6a4121f5b2a94acc88d62c07d8ed1cd34ed63/staging/src/k8s.io/api/admission/v1/types.go#L40))
+- Version conversion support (similar to admission webhook's MatchPolicy)
+- Access the old object (similar to [transition
+  rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language#transition-rules)
+  and oldObject in AdmissionRequest)
+- Configurability, as motivated above.
+
+There are also lots of additional capabilities (response message formatting,
+failure policies, type safety, advanced matching rules, ...) that will be discussed
+in detail further on in this proposal.
+
+We have divided this proposal into phases, all of which must be completed before
+this feature graduates to beta. Our goal is to size the phases so that each
+can be completed in a single Kubernetes release cycle.
+
+### Phase 1
+
+#### API Shape
+
+Before getting into all the individual fields and capabilities, let's look at the
+general "shape" of the API.
+
+This enhancement introduces a new `ValidatingAdmissionPolicy` kind.
+
+Each `ValidatingAdmissionPolicy` resource defines a admission control policy.
+The resource contains the CEL expressions to validate the admission policy and
+declares how the admission policy may be configured for use.
+
+For example:
+
+```yaml
+# Policy definition
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "replicalimit-policy.example.com"
+spec:
+  paramSource:
+    group: rules.example.com
+    kind: ReplicaLimit
+    version: v1
+  match:
+    resourceRules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["deployments"]
+  denyReason: Invalid
+  validations:
+    - name: max-replicas
+      expression: "object.spec.replicas <= params.maxReplicas"
+      message: "object.spec.replicas must be no greater than {1}"
+      messageArgs: ["params.maxReplicas"]
+      # ...other rule related fields here...
+```
+
+The `spec.paramSource` field of the `ValidatingAdmissionPolicy` specifies the
+kind of resources used to parameterize this policy. For this example, it is
+configured by `ReplicaLimit` custom resources. Note in this example how the CEL
+expression references to the parameters via the CEL `params` variable, e.g.
+`params.maxReplicas`.
+
+`spec.match` specifies what resources this policy is designed to validate. This
```

Comment:
Is this more like a selector described at the policy level rather than the instance? If so, you might be able to omit it by focusing such extensibility on the policy binding instead.

Reply:
Or give the two match clauses different names so it's clearer what is going on and how they interact. E.g. "matchResources" and "matchObjects"

Reply:
@TristonianJones this is a selector at the policy level. The primary goal here is to allow the policy definition author declare the conditions where their policy makes sense. If the policy author is writing a policy to validate the contents of Pod resources, they benefit by specifying that since they can then write CEL expressions that can safely assume that the object is the type they expect and that the object is not null (which can be the case if they don't establish with these match rules that the operation must be "CREATE" or "UPDATE" but not "DELETE").
 

@lavalamp I like the idea of distinguishing them, they are semantically different.  How about `matchConstraints` for policy definitions and `matchResources` for policy bindings?

Reply:
I've applied this. We can finalize exact field names in API review.